[lldb] Fix LLDBMemoryReader resolving GOT pointers incorrectly

augusto2112 opened the pull request:
The DYLD shared cache, as an optimization, can merge GOT pointers from multiple images into one location, and fix all relative offsets to point to the new updated location.  When reading metadata, LLDB tries, as an optimization, to read memory from local files instead of live memory. This means that the relative offset will point to the old location in the GOT section. In this case, LLDB needs to re-read the offset from live memory, to get the correct offset in live memory.

rdar://160837587

augusto2112 committed:
[lldb] Fix LLDBMemoryReader resolving GOT pointers incorrectly

The DYLD shared cache, as an optimization, can merge GOT pointers from
multiple images into one location, and fix all relative offsets to point
to the new updated location.  When reading metadata, LLDB tries,
as an optimization, to read memory from local files instead of live
memory. This means that the relative offset will point to the old
location in the GOT section. In this case, LLDB needs to re-read the
offset from live memory, to get the correct offset in live memory.

rdar://160837587

augusto2112 commented:
https://github.com/swiftlang/swift/pull/85091

@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/85091

@swift-ci test

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +316,150 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+            "{0:x} at offset {1:x}",
+            address.getRawAddress(), offset);
+
+  auto offset_address = address + offset;
```
Can you spell out the type here? It's not obvious from the context

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +316,150 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+  if (!readMetadataFromFileCacheEnabled())
+    return offset_address;
+
+  // Addresses in the process don't apply.
```
"don't apply" -> "can be read directly"?

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +316,150 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+
+  // This is only necessary on Apple platforms.
+  auto *obj_file = lldb_offset_address.GetModule()->GetObjectFile();
+  if (!obj_file || !obj_file->GetArchitecture().GetTriple().isAppleMachO())
```
isOSDarwin() is a more precise condition to find platforms with a shared cache.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +316,150 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+  if (!obj_file || !obj_file->GetArchitecture().GetTriple().isAppleMachO())
+    return offset_address;
+
+  auto section = lldb_offset_address.GetSection();
```
Can you spell out the types here?

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +316,150 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+  // The segment/section pair has to be:
+  //__DATA_CONST,__got, __AUTH_CONST,__auth_got or __AUTH_CONST,__auth_ptr.
+  bool is_data_const_got =
+      segment->GetName() == "__DATA_CONST" && section->GetName() == "__got";
```
This should be a virtual function — if we can avoid it we shouldn't hardcode the name/segment of the GOT here, but rather in ObjectFileMachO.h

augusto2112 replied:
I added a new commit with this code which I will upstream after this is merged

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +316,150 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+
+  // Relative offsets are always 4 bytes long, regardless of target.
+  uint32_t live_offset = 0;
+  auto size = sizeof(live_offset);
```
```suggestion
  size_t size = sizeof(live_offset);
```

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +316,150 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+  // Some Swift metadata encodes the directness directly into the offset,
+  // in that case clear the directness bit.
+  if (directnessEncodedInOffset)
+    live_offset &= ~1u;
```
is ~1u the right thing or should it be ~1ULL?

augusto2112 replied:
Since live_offset is always 4 bytes long ~1u always works, regardless of the platform.

augusto2112 committed:
[lldb] Add function to tell if a section is a GOT section

A global offset table is a section that holds the address of functions
that are dynamically linked.

rdar://160837587

augusto2112 commented:
https://github.com/swiftlang/swift/pull/85091

@swift-ci test

adrian-prantl commented on lldb/include/lldb/Core/Section.h:
```diff
@@ -278,6 +278,9 @@ class Section : public std::enable_shared_from_this<Section>,
   bool ContainsOnlyDebugInfo() const;
 
+  /// Returns true if this is a global offset table section.
+  bool IsGOTSection() const;
```
Please upstream this afterwards!

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +317,132 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+    return offset_address;
+  }
+
+  auto lldb_offset_address = *maybeAddr;
```
Can you add the type here?
LLVM coding style suggests only using auto when the type is altready on the same line (i.e., `auto var = dyn_cast<TypeName>()` or if it's excessively long.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -317,6 +317,132 @@ LLDBMemoryReader::resolvePointer(swift::remote::RemoteAddress address,
+
+  // Now, get the live address counterpart of the lldb address this function
+  // started with, and add the live offset we just read to it.
+  auto live_address = lldb_addr.GetLoadAddress(&target);
```
same here

augusto2112 commented:
https://github.com/swiftlang/swift/pull/85091

@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/85091

@swift-ci test
